game-core/src/test/java/games/strategy/triplea/delegate/power/calculator/TotalPowerAndTotalRollsTest.java

```diff
@@ -1,33 +1,690 @@
+      final Unit unit2 = givenUnit("test2", gameData);
+      unit2.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(-1);
+      final List<Unit> units = List.of(unit, unit2);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(9, 4, false)));
+      assertThat(
+          "The non infinite attack is used first",
+          sortedDie,
+          is(
+              List.of(
+                  new Die(1, 3, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT))));
+    }
+
+    @Test
+    void oneAaWithOverStack() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(2).setMayOverStackAa(true);
+      final List<Unit> units = List.of(unit);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(4, 2, true)));
+      assertThat(
+          sortedDie, is(List.of(new Die(1, 2, Die.DieType.HIT), new Die(1, 2, Die.DieType.HIT))));
+    }
+
+    @Test
+    void oneAaWithOverStackAndMoreRollsThanTargets() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(5).setMayOverStackAa(true);
+      final List<Unit> units = List.of(unit);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(10, 5, true)));
+      assertThat(
+          sortedDie,
+          is(
+              List.of(
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT),
+                  new Die(1, 2, Die.DieType.HIT))));
+    }
+
+    @Test
+    void oneAaWithOverstackAndInfinite() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(-1).setMayOverStackAa(true);
+      final List<Unit> units = List.of(unit);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(8, 4, true)));
```

Comment:
assertion comments giving the math behind the '8' and '4' I think would be of help. 
It might even make sense to assert one value at a time. That would help make it simpler per assertion and better adhere to the principle for assertions to have only one thing they are checking, one reason for them to ever fail. For example:
```
assertThat("4 rolls * 2 attack power = 8 total power", result.getFirst(), is(8));
```

Reply:
@trevan  bump ^